test/unit/contract_test.rb

```diff
@@ -100,16 +100,28 @@ class ContractTest < ActiveSupport::TestCase
   end
 
   test '.permitted_for' do
-    cinstances = FactoryBot.create_list(:simple_cinstance, 2)
-    user = FactoryBot.build(:member)
+    provider = FactoryBot.create(:simple_provider)
+    service1 = FactoryBot.create(:simple_service, account: provider)
+    service2 = FactoryBot.create(:simple_service, account: provider)
+    cinstances1 = FactoryBot.create_list(:simple_cinstance, 2, plan: FactoryBot.create(:simple_application_plan, issuer: service1))
+    cinstances2 = FactoryBot.create_list(:simple_cinstance, 2, plan: FactoryBot.create(:simple_application_plan, issuer: service2))
+    another_provider = FactoryBot.create(:simple_provider)
+    another_service = FactoryBot.create(:simple_service, account: another_provider)
+    FactoryBot.create_list(:simple_cinstance, 2, plan: FactoryBot.create(:simple_application_plan, issuer: another_service))
 
-    user.stubs(forbidden_some_services?: false)
-    permitted_contract_ids = Contract.permitted_for(user).pluck(:id)
-    cinstances.each { |contract| assert_includes(permitted_contract_ids, contract.id) }
+    user = FactoryBot.build(:member, account: provider)
 
-    user.stubs(forbidden_some_services?: true)
-    user.stubs(member_permission_service_ids: [cinstances.first.service_id])
-    assert_equal [cinstances.first.id], Contract.permitted_for(user).pluck(:id)
+    user.update(allowed_service_ids: nil, allowed_sections: ['plans'])
+    permitted_contract_ids = Cinstance.permitted_for(user).pluck(:id)
+    all_cinstances_of_provider = cinstances1.pluck(:id) + cinstances2.pluck(:id)
+
+    assert_same_elements all_cinstances_of_provider, permitted_contract_ids
```

Comment:
I rewrote this test to simulate a real behavior, and adding another service of the same provider, and another provider.

Before the `provider_for` scope fix, `permitted_contract_ids` returned contract IDs of another provider, because with permission status `:all`, `Contract.all` was returned.

Reply:
btw in a real reaquest the tenant leak detection should stop it but good to rely on correctness, not last resort safety net